### PR TITLE
Use StatsD format for metrics and sink to message channel

### PR DESCRIPTION
**Pull request by jkschneider**

### CODE REVIEW

File: spring-cloud-stream/pom.xml [hunk 1]
```diff
@@ -19,6 +19,11 @@
 			<artifactId>spring-boot-starter-actuator</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>io.micrometer</groupId>
+			<artifactId>micrometer-registry-statsd</artifactId>
```
Comment (jkschneider) [inline spring-cloud-stream/pom.xml]: Will have to document this somehow.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/ApplicationMetricsProperties.java [hunk 2]
```diff
@@ -119,12 +119,8 @@ public void setScheduleInterval(String scheduleInterval) {
 		this.scheduleInterval = scheduleInterval;
 	}
 
-	private boolean isMatch(String name, String[] includes, String[] excludes) {
-		if (ObjectUtils.isEmpty(includes)
-				|| PatternMatchUtils.simpleMatch(includes, name)) {
-			return !PatternMatchUtils.simpleMatch(excludes, name);
-		}
-		return false;
+	private boolean isMatch(String name, String[] includes) {
```
Comment (jkschneider) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/ApplicationMetricsProperties.java]: IDE suggested simplification.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/ApplicationMetricsProperties.java [hunk 3]
```diff
@@ -137,15 +133,12 @@ private boolean isMatch(String name, String[] includes, String[] excludes) {
...
+				if (isMatch(entry.getKey(), this.properties)) {
 					String stringValue = ObjectUtils.nullSafeToString(entry.getValue());
-					Object exportedValue = null;
-					if (stringValue != null) {
-						exportedValue = stringValue.startsWith("#{")
-								? beanExpressionResolver.evaluate(
-										environment.resolvePlaceholders(stringValue), expressionContext)
-								: environment.resolvePlaceholders(stringValue);
-					}
+					Object exportedValue = stringValue.startsWith("#{")
```
Comment (jkschneider) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/ApplicationMetricsProperties.java]: IDE suggested simplification

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java [hunk 4]
```diff
@@ -16,65 +16,75 @@
...
-			ApplicationMetricsProperties applicationMetricsProperties,
-			MetersPublisherBinding publisherBinding,
-			MetricsPublisherConfig metricsPublisherConfig, Clock clock) {
-		return new DefaultDestinationPublishingMeterRegistry(applicationMetricsProperties, publisherBinding, metricsPublisherConfig, clock);
+		StatsdMeterRegistry registry = StatsdMeterRegistry.builder(StatsdConfig.DEFAULT)
+				.clock(clock)
+				.lineSink(line -> publisherBinding.applicationMetrics().send(new GenericMessage<String>(line)))
+				.build();
+
+		registry.config().meterFilter(MeterFilter.denyUnless(id -> id.getName().startsWith("spring.integration")));
```
Comment (jkschneider) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java]: To cut down on noise, whitelist to those metrics that SCDF cares about.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java [hunk 5]
```diff
@@ -16,65 +16,75 @@
...
-	@Bean
-	@ConditionalOnMissingBean
-	public DefaultDestinationPublishingMeterRegistry defaultDestinationPublishingMeterRegistry(
-			ApplicationMetricsProperties applicationMetricsProperties,
-			MetersPublisherBinding publisherBinding,
-			MetricsPublisherConfig metricsPublisherConfig, Clock clock) {
-		return new DefaultDestinationPublishingMeterRegistry(applicationMetricsProperties, publisherBinding, metricsPublisherConfig, clock);
+		StatsdMeterRegistry registry = StatsdMeterRegistry.builder(StatsdConfig.DEFAULT)
+				.clock(clock)
+				.lineSink(line -> publisherBinding.applicationMetrics().send(new GenericMessage<String>(line)))
```
Comment (jkschneider) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java]: This is what sinks StatsD lines on the output channel.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java [hunk 6]
```diff
@@ -16,65 +16,75 @@
...
+				.build();
+
+		registry.config().meterFilter(MeterFilter.denyUnless(id -> id.getName().startsWith("spring.integration")));
+
+		Collection<Tag> tags = applicationMetricsProperties.getExportProperties().entrySet().stream()
```
Comment (jkschneider) [inline spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java]: Convert to common tags that will be shipped with every metric.

File: spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java (hunk 5 above)
Comment (olegz) [reply spring-cloud-stream/src/main/java/org/springframework/cloud/stream/micrometer/MessageChannelStatsdMetricsAutoConfiguration.java]: John, i'll probably make consumer a bean with Conditional so users can override it

**Comment by olegz** [comment]
@jkschneider as discussed, we're going to put this one aside for now and instead especially after making few modification that allowed me bring the message format to what it was before. So the string argument here is backward compatibility. 
That said, we will definitely consider bringing `statsd` publisher for 2.1. I'll raise a separate issue and link to this PR

**Pull request closed without merging**